### DISCUSSION TITLE
Convert pathname

### DISCUSSION THREAD

**Discussion by MathiasSvDK** [amannn/next-intl#491] [Q&A]
### Is your feature request related to a problem? Please describe.

Not really a problem

### Describe the solution you'd like

It would be great for SEO to be able to convert the current path from the default langauge to xx langauge the user has.
It would be very much needed for <link rel="canonical"

### Describe alternatives you've considered

None

**Comment by MathiasSvDK** [comment]
What i request is like in localized pathname:
https://next-intl-docs.vercel.app/docs/routing/navigation
Be able to retrieve the current localized path from the URL into the html to be able to create the <link rel="canonical"

**Comment by amannn** [comment]
Can you elaborate on what kind of canonical link you have to render on your page and why it is necessary?

Note that [the `next-intl` middleware sets the `link` header](https://next-intl-docs.vercel.app/docs/routing/middleware) for you that provides hints to search engines about different language versions of your page.

**Comment by MathiasSvDK** [comment]
We have an issue with google thinking that our .com domain is a copy of .dk (That they don't "work" together).
The pages are identical but translated and we need to let google know that they are the same.

Maybe i am doing it wrong? Do you have any knowledge about this?

**Comment by amannn** [comment]
Do you have `alternateLinks` enabled in the `next-intl` middleware?

Please refer to [the docs](https://next-intl-docs.vercel.app/docs/routing/middleware#disable-alternate-links), there's also a link to Google documentation that explains how the `link` header works.

**Comment by MathiasSvDK** [comment]
I don't have it disabled.
I found in the docs, as you linked that it says "Defaults to `true`" - so i assume it's enabled by default?

**Comment by amannn** [comment]
Yes.

**Comment by MathiasSvDK** [comment]
Any suggestion to how i then can fix this issue?

**Comment by amannn** [comment]
Sorry, I'm not sure what the issue is. As mentioned, the middleware returns alternate links, therefore there's no need for canonical links (which shouldn't be used for language versions anyway).

I'll convert this to a discussion since it's a usage question, I hope you understand.

**Comment by Stinobe** [comment]
I'm having the same issue I think. I'm building a website with following slugs:

```javascript
{
  // ...
  "/services/[slug]": {
    "en": "/services/[slug]",
    "nl": "/diensten/[slug]"
  },
  // ...
}
```

For example in the Dutch headers I have
```text
</en/services/levering>; rel="alternate"; hreflang="en",
</nl/diensten/levering>; rel="alternate"; hreflang="nl",
</diensten/levering>; rel="alternate"; hreflang="x-default"
```
The English link will give a 404 because it should be `/en/services/delivery`.

I don't see any other solution than disabling the `alternates` tbh since there's no way to know the translation of the dynamic slug.
At the moment I'm working with the `getPathname` function and trying to automate it as much as possible.

**Reply by Stinobe** [reply]
So here's what I came up with. It would be helpful if the property `params` could be optional / undefined for non-dynamic paths instead of omitted. I couldn't get the types to match so I had to infer it.

```typescript
// Get whatever I have defined as languages & pathnames
import { i18nLanguages, i18nLocales, AppPathnames } from "./i18n";
// Use getPathname fn
import { getPathname } from "./navigation";

// Extract the href parameter type from getPathname fn
type HrefParams = Exclude<
  Parameters<typeof getPathname>[0]["href"],
  AppPathnames
>;

// Get the parameters for a given path
type hrefWithParams<T extends AppPathnames> = Extract<
  HrefParams,
  { pathname: T }
>;

// Get types for params & query parameters
// Some logic is added because params isn't available for non-dynamic paths
type TranslatedPathnameParams<T extends AppPathnames> =
  hrefWithParams<T> extends { params: any }
    ? Pick<hrefWithParams<T>, "query" | "params">
    : Pick<hrefWithParams<T> & { params?: undefined }, "query" | "params">;

// To be passed through to Metadata.alternates
type PathInfo = { canonical?: string; languages: { [x: string]: string } };

// Create dynamic fn props
type DynamicParams<T extends AppPathnames> = {
  params: {
    [x in i18nLanguages]: TranslatedPathnameParams<T>["params"];
  };
  query?: { [x in i18nLanguages]?: TranslatedPathnameParams<T>["query"] };
};

// The functionality itself
export const getTranslatedPathnames = <T extends AppPathnames>(
  locale: i18nLanguages,
  pathname: T,
  params?: DynamicParams<T>["params"],
  queries?: DynamicParams<T>["query"],
): PathInfo => {
  const alternates: PathInfo = { languages: {} };

  // Loop over each language
  i18nLocales.forEach((lang) => {
    // Get query for current language
    const query = queries && queries[lang];
    // Get parameters for current language
    const param = params && params[lang];
    // Create pathinfo
    const pathInfo = getPathname({
      locale: lang,
      href: (params
        ? { pathname, query, params: param }
        : { pathname, query }) as HrefParams, // Just being a bit to sure about myself
    });

    // If current language is same as current locale, set it as canonical url
    if (lang === locale) alternates.canonical = `/${lang}${pathInfo}`;
    // Otherwise set it as alternate url
    else {
      if (!alternates.languages) alternates.languages = {};
      alternates.languages[lang] = `/${lang}${pathInfo}`;
    }
  });

  // Return the object to be spread in metadata
  return alternates;
};

```
